File: nextframe-cli/src/cli/compose.js
import { spawnSync } from "node:child_process";
import { mkdir, stat } from "node:fs/promises";
import { dirname, resolve as resolvePath } from "node:path";
import { fileURLToPath } from "node:url";

import { parseFlags, emit } from "./_io.js";
import { resolveTimeline, timelineUsage } from "./_resolve.js";

const USAGE = timelineUsage("compose", "", " [out.html]");
const BUNDLE_PATH = resolvePath(
  dirname(fileURLToPath(import.meta.url)),
  "../../../runtime/web/src/bundle.cjs"
);

function defaultOutputPath(jsonPath) {
  return jsonPath.endsWith(".json") ? `${jsonPath.slice(0, -5)}.html` : `${jsonPath}.html`;
}

export async function run(argv) {
  const { positional, flags } = parseFlags(argv);
  const resolved = resolveTimeline(positional, { usage: USAGE });
  if (!resolved.ok) {
    emit(resolved, flags);
    return resolved.error?.code === "USAGE" ? 3 : 2;
  }

  if (!resolved.legacy && resolved.rest.length > 0) {
    emit({ ok: false, error: { code: "USAGE", message: USAGE } }, flags);
    return 3;
  }

  if (resolved.legacy && resolved.rest.length > 1) {
    emit({ ok: false, error: { code: "USAGE", message: USAGE } }, flags);
    return 3;
  }

  const outputPath = resolved.legacy
    ? resolvePath(resolved.rest[0] || defaultOutputPath(resolved.jsonPath))
    : defaultOutputPath(resolved.jsonPath);

  await mkdir(dirname(outputPath), { recursive: true });

  const result = spawnSync("node", [BUNDLE_PATH, resolved.jsonPath, outputPath], {
    encoding: "utf8",
    stdio: ["ignore", "pipe", "pipe"],
  });

  if (result.error) {
    emit({
      ok: false,
      error: {
        code: "COMPOSE_SPAWN",
        message: `failed to spawn node bundle.cjs: ${result.error.message}`,
      },
    }, flags);
    return 2;
  }

  if (result.status !== 0) {
    const detail = (result.stderr || result.stdout || "").trim();
    emit({
      ok: false,
      error: {
        code: "COMPOSE_FAILED",
        message: detail || `bundle.cjs exited with code ${result.status ?? -1}`,
      },
    }, flags);
    return 2;
  }

  const outputStat = await stat(outputPath);
  emit({
    ok: true,
    value: {
      path: outputPath,
      size: outputStat.size,
    },
  }, flags);
  return 0;
}

File: bridge/src/compose.rs
use serde_json::{json, Value};
use std::fs;
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};

use crate::validation::require_string;

fn replace_json_extension(path: &Path) -> PathBuf {
    let as_string = path.to_string_lossy();
    if let Some(stripped) = as_string.strip_suffix(".json") {
        return PathBuf::from(format!("{stripped}.html"));
    }
    PathBuf::from(format!("{as_string}.html"))
}

pub(crate) fn handle_compose_generate(params: &Value) -> Result<Value, String> {
    let timeline_path = PathBuf::from(require_string(params, "timelinePath")?);
    let output_path = match params.get("outputPath").and_then(Value::as_str) {
        Some(raw) if !raw.trim().is_empty() => PathBuf::from(raw),
        _ => replace_json_extension(&timeline_path),
    };

    let bundle_path = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../runtime/web/src/bundle.cjs");

    let output = Command::new("node")
        .arg(&bundle_path)
        .arg(&timeline_path)
        .arg(&output_path)
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .map_err(|error| format!("failed to run compose bundle: {error}"))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr).trim().to_string();
        let stdout = String::from_utf8_lossy(&output.stdout).trim().to_string();
        let details = if !stderr.is_empty() {
            stderr
        } else if !stdout.is_empty() {
            stdout
        } else {
            format!("bundle exited with code {}", output.status.code().unwrap_or(-1))
        };
        return Err(details);
    }

    let meta = fs::metadata(&output_path)
        .map_err(|error| format!("failed to stat composed html {}: {error}", output_path.display()))?;

    Ok(json!({
        "ok": true,
        "path": output_path,
        "size": meta.len(),
    }))
}
